### PR TITLE
Module 4 - Filter on activity page

### PR DESCRIPTION
**Pull request by sinpai**
* Added filter on activity page by:
 - Date
 - From balance to selected category
 - From one selected category to another

Screen - http://take.ms/U9kij
Video - http://take.ms/BEDtO

### CODE REVIEW

File: app/views/transactions/_search.html.slim [hunk 1]
```diff
@@ -1,8 +1,32 @@
...
+      .row
+        .col-md-12
+          h6
+            = t('.filter_by_category')
+    .card-body.d-flex.flex-row.justify-content-around.align-items-center.py-2
+      div
+        h6
+          =t('.select_date_range')
+        div.d-flex.flex-column
+          = f.date_field :transactinable_of_BalanceTransaction_type_date_gt
```
Comment (AleksSenkou) [inline app/views/transactions/_search.html.slim]: @sinpai can't we use snake case? `transactinable_of_balance_transaction_type_date_gt`

File: app/views/transactions/_search.html.slim [hunk 2]
```diff
@@ -1,8 +1,32 @@
...
+            = t('.for_transaction_to_category')
+          = f.select :transactinable_of_CategoryTransaction_type_category_id_in,
+            options_for_select(current_user.categories.collect { | category | [category.name, category.id] }) || [], include_blank: true
+        div.mt-2
+          h6
+            = t('.for_transactions_between_categories')
+          .row
+            .col
+              = f.select :transactinable_of_CrossCategoriesTransaction_type_category_from_id_in,
+              options_for_select(current_user.categories.collect { | category | [category.name, category.id] }) || [], include_blank: true
```
Comment (AleksSenkou) [inline app/views/transactions/_search.html.slim]: @sinpai `|category|` instead of `| category |`

File: app/views/transactions/_search.html.slim [hunk 3]
```diff
@@ -1,8 +1,32 @@
...
+          =t('.select_date_range')
+        div.d-flex.flex-column
+          = f.date_field :transactinable_of_BalanceTransaction_type_date_gt
+          = f.date_field :transactinable_of_BalanceTransaction_type_date_lt
+      div.d-flex.flex-column
+        div
+          h6
+            = t('.for_transaction_to_category')
+          = f.select :transactinable_of_CategoryTransaction_type_category_id_in,
+            options_for_select(current_user.categories.collect { | category | [category.name, category.id] }) || [], include_blank: true
```
Comment (AleksSenkou) [inline app/views/transactions/_search.html.slim]: @sinpai 
```ruby
options_for_select(current_user.categories.collect { | category | [category.name, category.id] }) || [], include_blank: true
```
is duplicated. Let's do with DRY and place these code in helper

File: config/routes.rb [hunk 4]
```diff
@@ -3,7 +3,8 @@
   root to: 'home_page#show'
   resources :transactions, only: %i[index] do
     collection do
-      post 'search', to: 'transactions#search'
+      post 'search' => 'transactions#search'
+      get 'search' => 'transactions#index'
```
Comment (AleksSenkou) [inline config/routes.rb]: @sinpai we can use `match` here
```ruby
match 'search', to: 'transactions#search', via: %i[get post], as: :search
```

**Review by AleksSenkou — COMMENTED** [review]
Suggestions:

File: app/views/transactions/_search.html.slim (hunk 1 above)
Comment (sinpai) [reply app/views/transactions/_search.html.slim]: In practice, snake case isn't applicable here. Ransack matcher doesn't work after changes.

File: app/helpers/transactions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+module TransactionsHelper
+  def category_options
+    options_for_select(current_user.categories.map { |category| [category.name, category.id] }) || []
+  end
```
Comment (AleksSenkou) [inline app/helpers/transactions_helper.rb]: @sinpai looks like a huge line. Let's split it too several:
```ruby
def category_options
  options_for_select(categories_by_name_and_id) || []
end

private

def categories_by_name_and_id
  current_user.categories.map { |category| [category.name, category.id] }
end
```

Comment (AleksSenkou) [reply app/helpers/transactions_helper.rb]: @sinpai why do we need `[]` here? wouldn't `options_for_select([])` return empty array?

File: app/views/transactions/_search.html.slim [hunk 6]
```diff
@@ -1,8 +1,32 @@
...
+            = t('.for_transaction_to_category')
+          = f.select :transactinable_of_CategoryTransaction_type_category_id_in,
+            category_options, include_blank: true
+        div.mt-2
+          h6
+            = t('.for_transactions_between_categories')
+          .row
+            .col
+              = f.select :transactinable_of_CrossCategoriesTransaction_type_category_from_id_in,
+              category_options, include_blank: true
```
Comment (AleksSenkou) [inline app/views/transactions/_search.html.slim]:
```ruby
= f.select :transactinable_of_CrossCategoriesTransaction_type_category_from_id_in,
  category_options, include_blank: true
```

**Review by AleksSenkou — COMMENTED** [review]
Suggestions

**Review by AleksSenkou — APPROVED** [review]
LGTM

**Pull request merged**